mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java

```diff
@@ -0,0 +1,239 @@
+        }
+
+        @FunctionalInterface
+        interface RequireDeletionDate<T> {
+            T deletionDate(ZonedDateTime deletionDate);
+        }
+
+        @FunctionalInterface
+        interface RequireSender<T> {
+            T sender(MaybeSender sender);
+        }
+
+        @FunctionalInterface
+        interface RequireRecipients<T> {
+            T recipients(List<MailAddress> recipients);
+
+            default T recipients(MailAddress... recipients) {
+                return recipients(ImmutableList.copyOf(recipients));
+            }
+        }
+
+        @FunctionalInterface
+        interface RequireHasAttachment<T> {
+            T hasAttachment(boolean value);
+
+            default T hasAttachment() {
+                return hasAttachment(true);
+            }
+
+            default T hasNoAttachments() {
+                return hasAttachment(false);
+            }
+        }
+
+        @FunctionalInterface
+        interface RequireContent<T> {
+            T content(Supplier<InputStream> content);
+        }
+
+        public static class FinalStage {
+            private final MessageId messageId;
+            private final List<MailboxId> originMailboxes;
+            private final User owner;
+            private final ZonedDateTime deliveryDate;
+            private final ZonedDateTime deletionDate;
+            private final MaybeSender sender;
+            private final List<MailAddress> recipients;
+            private final boolean hasAttachment;
+            private final Supplier<InputStream> content;
+            private Optional<String> subject;
+
+            public FinalStage(MessageId messageId, List<MailboxId> originMailboxes, User owner, ZonedDateTime deliveryDate,
+                              ZonedDateTime deletionDate, MaybeSender sender, List<MailAddress> recipients, boolean hasAttachment,
+                              Supplier<InputStream> content) {
+                this.messageId = messageId;
+                this.originMailboxes = originMailboxes;
+                this.owner = owner;
+                this.deliveryDate = deliveryDate;
+                this.deletionDate = deletionDate;
+                this.sender = sender;
+                this.recipients = recipients;
+                this.hasAttachment = hasAttachment;
+                this.content = content;
+                this.subject = Optional.empty();
+            }
+
+            public FinalStage subject(String subject) {
+                this.subject = Optional.of(subject);
+                return this;
+            }
+
+            public DeletedMessage build() {
+                return new DeletedMessage(messageId, originMailboxes, owner, content, deliveryDate, deletionDate, sender,
+                    recipients, subject, hasAttachment);
+            }
+        }
+    }
+
+    public static Builder.RequireMessageId<Builder.RequireOriginMailboxes<Builder.RequireUser<Builder.RequireDeliveryDate<
```

Comment:
We should find a better to present it.
The closing greater than symbols look like a git conflict ;)

Reply:
I added some intermediary steps to the design...